pyomo/contrib/mpc/modeling/cost_expressions.py

```
@@ -246,3 +277,52 @@ def tracking_rule(m, t):
         return sum(cost[t] for cost in tracking_costs)
     tracking_cost = Expression(time, rule=tracking_rule)
     return tracking_cost
+
+
+def get_constraint_residual_expression(
+    constraints,
+    time,
+    weight_data=None,
+    # TODO: Option for norm (including no norm)
+):
+    if weight_data is None:
+        weight_data = ScalarData(
+            ComponentMap((var, 1.0) for con in constraints)
+        )
+    if not isinstance(weight_data, ScalarData):
+        weight_data = ScalarData(weight_data)
+    for con in constraints:
+        if not weight_data.contains_key(con):
+            raise KeyError(
+                "Tracking weight does not contain a key for"
+                " constraint %s" % con
+            )
+    n_con = len(constraints)
+    con_set = Set(initialize=range(n_con))
+    resid_expr_list = []
+    for con in constraints:
+        resid_expr_dict = {}
+        for t in time:
+            expr = con[t].expr
+            if isinstance(expr, EqualityExpression):
+                resid_expr_dict[t] = (con[t].body - con[t].upper)
+            elif con.upper is None:
+                resid_expr_dict[t] = (con[t].lower - con[t].body)
+            elif con.lower is None:
+                resid_expr_dict[t] = (con[t].body - con[t].upper)
```

Do we need these cases? I thought we should raise the error if the model constraints are inequalities.